content/post/postgres/rfc-extension-packaging-lookup.md

```diff
@@ -0,0 +1,496 @@
+
+Then modify the preloading code to also preload these files. For each
+extension in a list, it would:
+
+*   Search each `$extension_path` for the extension.
+*   When found, load all the shared libraries in `$extension/lib`.
+
+For example, to load all shared modules in the `pg_partman` extension, set:
+
+```ini
+shared_preload_extensions = 'pg_partman'
+```
+
+To load a single shared module from an extension, give its name after the
+extension name and a slash. This example will load only the `pg_partman_bgw`
+shared module from the `pg_partman` extension:
+
+```ini
+shared_preload_extensions = 'pg_partman/pg_partman_bgw'
+```
+
+This change requires a one-time change to existing preload configurations on
+upgrade.
+
+## Future: Deprecate LOAD
+
+For a future change, consider modifying `CREATE EXTENSION` to support shared
+module-only extensions. This would allow extensions with no SQL component,
+such as `auto_explain` to be handled like any other extension; it would live
+in `--extdir-core` with a directory structure like this:
+
+``` tree
+auto_explain
+├── auto_explain.control
+└── lib
+   ├── auto_explain.dylib
+   └── bitcode
+       ├── auto_explain
+       │   └── auto_explain.bc
+       └── auto_explain.index.bc
+```
+
+Note the `auto_explain.control` file. We would need to add a variable to
+indicate that the extension includes no SQL files, so `CREATE EXTENSION` and
+related commands wouldn't try to find them.
+
+With these changes, extensions could become the primary, recommended interface
+for extending PostgreSQL. Perhaps the `LOAD` command could be deprecated, and
+the `*_preload_libraries` GUCs along with it.
+
+## Compatibility Issues
+
+*   The `directory` and `module_pathname` control file variables and the
+    `MODULEDIR` PGXS variable would be deprecated and ignored.
+*   `*_preload_libraries` would no longer be used to find extension modules
+    without full paths. Administrators would have to remove module names from
+    these GUCs and add the relevant extension names to the
+    `*_preload_extensions` variables. To ease upgrades, we might consider
+    adding a PGXS variable that, when true, would symlink shared modules into
+    `--pkglibdr`.
+*   `LOAD` would no longer be able to find shared modules included with
+    extensions, unless we add a PGXS variable that, when true, would symlink
+    shared modules into `--pkglibdr`.
+*   The `EXTENSION` PGXS variable will no longer support multiple extension
+    names.
+*   The change in extension installation locations must also be adopted by
+    projects that don't use PGXS for installation, like [pgrx]. Or perhaps
+    they could be modified to also use PGXS. Long term it might be useful to
+    replace the `Makefile`-based PGXS with another installation system,
+    perhaps a CLI.
+
+## Out of Scope
+
+This RFC does not include or attempt to address the following issues:
+
+*   How to manage third-party shared libraries. Making system dependencies
```

Comment:
This is kind of an elephant in the room, and I definitely believe this RFC would be the place to address it --or else it won't be addressed again until a new further change to PG core for extensions, which may be a very distant future.

Reply:
Me too, just somebody has to do the research, work out a proposal, and write it down. But in the absence of that (and yes, I will ask around for contributions), I want to work on incremental change, so that the perfect isn't the enemy of the good, and because that's how things are often done in the project.

Hell, Peter E submitted [a patch](https://postgr.es/m/0d384836-7e6e-4932-af3b-8dad1f6fee43@eisentraut.org) proposing just the extension path stuff and no reorganization at all. As long as it's a step in the right direction I'm all in favor.

Reply:
Fair enough!

Reply:
That said, maybe we could borrow something from [Conda](https://docs.conda.io/projects/conda-build/en/latest/resources/use-shared-libraries.html)?

Reply:
Specifically this bit:

> Conda-build uses @loader_path on macOS and $ORIGIN on Linux because we install into a common root directory and can assume that other libraries are also installed into that root. The use of the variables allows you to build relocatable binaries that can be built on one system and sent everywhere.

Which I'll be looking into when I get to building binary packages.

Reply:
Maybe, sure. What I take it from there is that dependencies can reside on a `lib/` subfolder relative to the extension. And this idea is already part of the RFC.

Reply:
Well this RFC's `lib` folder is for `pkglibdir` files, but yeah.